doc/arch/014-reporting-etl-pipeline.md

```diff
@@ -0,0 +1,150 @@
+# Reporting ETL Pipeline
+
+## Context
+
+Simple's dashboard reports are currently generated by queries on the application database's tables. This has several
+disadvantages as we start to scale.
+
+### 1. Complexity
+
+Simple's application schema is optimized for application transactions like registering patients, recording BPs,
+updating patient info, or tracking overdue patients. As such, reporting queries grow complex. Asking questions like
+"latest BP in a month", "prescribed drugs as of a date", "current step of treatment protocol" involve complex clauses
+in the SQL queries.
+
+This complexity has a tangible impact on the quality of our reports. We risk data inconsistency as complex queries are
+more error-prone when reused in multiple places.
+
+
+### 2. Coupling
+
+When the application schema is used for reporting also, there is a tight coupling between application requirements and
+reporting requirements. Changes to the schema to optimize for reporting may have a negative impact on complexity in the
+application, and vice versa.
+
+In addition to complexity, performance is also a growing concern as we scale. Heavy reporting operations may strain the
+application's ability to write data to the database. Conversely, heavy application activity may impact our ability to
+generate reports in a performant manner. We currently use heavy layers of caching to mitigate these performance
+concerns, but the less caching we _need_ to do, the better.
+
+### Priorities and constraints
+
+The upcoming changes to our reporting pipeline are quite new to us, so the priority is the ability to iterate.
```

Comment:
The more I think about the "new" ETL we are talking about, the more I realize it isn't really new...its mostly a refined, better approach to the materialized views we already have in the app and have been using for a long time now. So in some ways, its already an iteration on a previous approach that has worked out well. That may be worth mentioning here...we aren't switching to some brand new ETL architecture in a separate database, we are instead working towards reporting views in our existing DB that are modeled better and more directly useful for the reports we need today.